core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,503 @@
+    for {
+      executor <- (1 to 4).map(_.toString)
+      index <- 0 until 10
+    } {
+      withClue(s"exec = $executor; index = $index") {
+        val badExec = (executor == "1" || executor == "2")
+        val badPart = (index == 0 || index == 1)
+        val taskExp = (badExec && badPart)
+        assert(
+          tsm.isExecutorBlacklistedForTask(executor, index) === taskExp)
+        val executorExp = badExec
+        assert(tsm.isExecutorBlacklistedForTaskSet(executor) === executorExp)
+      }
+    }
+    assertEquivalentToSet(tsm.isNodeBlacklistedForTaskSet, Set("hostA"))
+    // we dont' blacklist the nodes or executors till the stages complete
+    assert(blacklistTracker.nodeBlacklist() === Set())
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set())
+    assertEquivalentToSet(blacklistTracker.isExecutorBlacklisted(_), Set())
+
+    // when the stage completes successfully, now there is sufficient evidence we've got
+    // bad executors and node
+    blacklistTracker.updateBlacklistForSuccessfulTaskSet(0, 0, tsm.execToFailures)
+    assert(blacklistTracker.nodeBlacklist() === Set("hostA"))
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set("hostA"))
+    assertEquivalentToSet(blacklistTracker.isExecutorBlacklisted(_), Set("1", "2"))
+
+    clock.advance(blacklistTracker.BLACKLIST_TIMEOUT_MILLIS + 1)
+    blacklistTracker.applyBlacklistTimeout()
+    assert(blacklistTracker.nodeBlacklist() === Set())
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set())
+    assertEquivalentToSet(blacklistTracker.isExecutorBlacklisted(_), Set())
+  }
+
+  def trackerFixture: (BlacklistTracker, TaskSchedulerImpl) = {
+    trackerFixture()
+  }
+
+  def trackerFixture(confs: (String, String)*): (BlacklistTracker, TaskSchedulerImpl) = {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set(config.BLACKLIST_ENABLED.key, "true")
+    confs.foreach { case (k, v) => conf.set(k, v) }
+    val scheduler = mockTaskSchedWithConf(conf)
+
+    clock.setTime(0)
+    blacklistTracker = new BlacklistTracker(conf, clock)
+    (blacklistTracker, scheduler)
+  }
+
+  test("executors can be blacklisted with only a few failures per stage") {
+    val (tracker, scheduler) = trackerFixture
+    // for 4 different stages, executor 1 fails a task, then executor 2 succeeds the task,
+    // and then the task set is done.  Not enough failures to blacklist the executor *within*
+    // any particular taskset, but we still blacklist the executor overall eventually
+    (0 until 4).foreach { stage =>
+      val taskSet = FakeTask.createTaskSet(1)
+      val tsm = new TaskSetManager(scheduler, Some(tracker), taskSet, 4, clock)
+      tsm.updateBlacklistForFailedTask("hostA", "1", 0)
+      tracker.updateBlacklistForSuccessfulTaskSet(stage, 0, tsm.execToFailures)
+    }
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set("1"))
```

Comment:
like above,  can you add incremental assertions in the for-loop to make sure that the executor is *not* blacklisted until enough things fail?